twine/package.py

```diff
@@ -184,3 +164,66 @@ def sign(self, sign_with, identity):
         subprocess.check_call(gpg_args)
 
         self.add_gpg_signature(self.signed_filename, self.signed_basefilename)
+
+
+class HashManager(object):
+    """Manage our hashing objects for simplicity.
+
+    This will also allow us to better test this logic.
+    """
+
+    def __init__(self, filename):
+        """Initialize our manager and hasher objects."""
+        self.filename = filename
+        try:
+            self._md5_hasher = hashlib.md5()
+        except ValueError:
+            # FIPs mode disables MD5
+            self._md5_hasher = None
+
+        self._sha2_hasher = hashlib.sha256()
+        self._blake_hasher = None
+        if blake2b is not None:
+            self._blake_hasher = blake2b(digest_size=256 // 8)
+
+    def _md5_update(self, content):
+        if self._md5_hasher is not None:
+            self._md5_hasher.update(content)
+
+    def _md5_hexdigest(self):
+        if self._md5_hasher is not None:
+            return self._md5_hasher.hexdigest()
+        return None
+
+    def _sha2_update(self, content):
+        if self._sha2_hasher is not None:
+            self._sha2_hasher.update(content)
+
+    def _sha2_hexdigest(self):
+        if self._sha2_hasher is not None:
+            return self._sha2_hasher.hexdigest()
+        return None
+
+    def _blake_update(self, content):
+        if self._blake_hasher is not None:
+            self._blake_hasher.update(content)
+
+    def _blake_hexdigest(self):
+        if self._blake_hasher is not None:
+            return self._blake_hasher.hexdigest()
+        return None
+
+    def hash(self):
```

Comment:
This could get called automagically, or we could add a function that creates the object, calls this method, and then returns the hexdigest, but I wasn't sure that was necessary either